lib/kinesis_client/stream/shard/producer.ex

```diff
@@ -248,6 +248,23 @@ defmodule KinesisClient.Stream.Shard.Producer do
             message:
               "Checkpoint failed for #{state.app_name}-#{state.shard_id}: this consumer is not the lease owner"
 
+        {:error, {"ProvisionedThroughputExceededException", _}} ->
+          if attempt > 10 do
+            raise KinesisClient.Error,
+              message:
+                "Checkpoint failed for #{state.app_name}-#{state.shard_id}: provisioned throughput exceeded after 10 attempts"
+          end
+
+          :telemetry.execute(
+            [:kinesis_client, :shard_producer, :checkpoint_error],
+            %{reason: :provisioned_throughput_exceeded, attempt: attempt},
+            telemetry_meta(state)
+          )
```

Comment:
While we're here I think it might be useful to modify this Telemetry event to emit whenever we raise an exception as well. And then replace `attempt` with a `fatal: true | false` measurement. We can then very easily filter for fatal events we know have crashed the consumer in Datadog graphs and monitors.

Reply:
Sounds good. I did think that the `span` this whole block is wrapped in gives us enough indicators for when things fail that way, but this would be more convenient.